pinc/project_quick_check.inc

```diff
@@ -323,19 +323,24 @@ function _test_project_for_bad_bytes($projectid)
             <th>" . _("Text") . "</th>
             <th>" . _("#") . "</th>
             <th>" . _("Raw") . "</th>
+            <th>" . _("Codepoints") . "</th>
             <th>" . _("Bytes") . "</th>
             <th>" . _("Likely intended character") . "</th>
             <th>" . _("Why bad") . "</th>
         </tr>
     ";
 
     $n_bad_pages = 0;
+    $valid_codepoints = $project->get_valid_codepoints();
+
     while (list($text, $imagename, $proofers) = page_info_fetch($pages_res))
     {
         $round_num = get_round_num_from_proofers($proofers);
-        $occurrences = find_bad_byte_sequences_in_text($text, $projectid);
+        $occurrences = find_bad_byte_sequences_in_text($text);
+        $invalid_chars = get_invalid_characters($text, $valid_codepoints);
```

Comment:
I understand that you want to output characters different than bytes. Can you explain why you needed to split this out to do that rather than change `tds_for_bad_bytes()` to output different data for characters vs other bad bytes? Doing it this way neuters `find_bad_byte_sequences_in_text()`.

Reply:
The problem was that `find_bad_byte_sequences_in_text()` just returns the strings which you then have to interpret in `tds_for_bad_bytes()` to work out what the problem was. An ambiguity can arise with a string starting with '&' which could in principle (though unlikely) be a combining sequence (when that part gets written) rather than a html entity so in the combining sequence branch I had to put in a 'hack' to distinguish that case. The former code used the `elseif(utf8_strlen($raw) == 1 && is_utf8($raw))`  line to distinguish invalid codepoints which also needed further adjustment to distinguish invalid combining characters.
Dealing with invalid characters and bad sequences separately avoids these problems.
Another approach would be to record more information in the '$occurrences' array so we did not  have to re-interpret the results but this would probably be more complicated.